### PR TITLE
feat: added data residency support

### PR DESCRIPTION
**Pull request by divyasreedev**
## Description of the change

- Added support for data residency feature
- Updated dart SDK version constraint to >=2.17.6 across all packages to use enhanced enums (we will need to support dart version > 3 as well in future
- Removed un-used dependencies across all packages
- Added Unit Tests Cases for the Data Residency Feature
- Upgraded to the latest version of the underlying Native SDKs

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Development

- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review 

- [x]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [x] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [x] Changes have been reviewed by at least one other engineer
- [x] Issue from task tracker has a link to this pull request

### CODE REVIEW

**Review by desusai7 — APPROVED** [review]

File: packages/example/ios/Podfile.lock [hunk 1]
```diff
@@ -20,94 +20,20 @@ PODS:
...
-    - GoogleUtilities/Logger
   - Leanplum-iOS-SDK (4.0.0)
-  - nanopb (2.30909.0):
-    - nanopb/decode (= 2.30909.0)
-    - nanopb/encode (= 2.30909.0)
-  - nanopb/decode (2.30909.0)
-  - nanopb/encode (2.30909.0)
-  - PromisesObjC (2.2.0)
-  - Rudder (1.11.0)
+  - Rudder (1.16.0)
```
Comment (bardisg) [inline packages/example/ios/Podfile.lock:36]: use latest afterrecent release if not used already

File: packages/example/ios/Podfile.lock [hunk 2]
```diff
@@ -225,36 +132,26 @@ SPEC CHECKSUMS:
...
-  nanopb: b552cce312b6c8484180ef47159bc0f65a1f0431
-  PromisesObjC: 09985d6d70fbe7878040aa746d78236e6946d2ef
-  Rudder: 9bb3913f997d2b248a7fd44c768994546c53411d
+  Rudder: 718bc5bf0de0e4d83850dedd792f08e5e2065a0a
   Rudder-Adjust: 5f14011c8f7237d80a96a10655ac03ebf832bcc4
   Rudder-Amplitude: f845cc125a1a58d4de6155391a2b0392815ae898
   Rudder-AppCenter: 9eca9241e3707a0e9610714dd91dc8da4bae7e1f
   Rudder-Appsflyer: b19834ae8d933444411813556e0fe70a33df224a
   Rudder-Braze: e58e0e8163a2dcf0d45ec2d12bdb0ec3329c3681
-  Rudder-Firebase: 9f061bf3c23900e1a8f32f8b079ae17e04874f17
```
Comment (bardisg) [inline packages/example/ios/Podfile.lock:246]: are we ok with this being removed?

File: packages/plugins/rudder_plugin_android/android/build.gradle [hunk 3]
```diff
@@ -38,6 +38,6 @@ android {
 }
 
 dependencies {
-    implementation 'com.rudderstack.android.sdk:core:[1.8.1, 1.9.0)'
+    implementation 'com.rudderstack.android.sdk:core:[1.14.0, 1.15.0)'
```
Comment (bardisg) [inline packages/plugins/rudder_plugin_android/android/build.gradle:41]: is this latest after recent relase?

File: packages/plugins/rudder_plugin_android/android/src/main/java/com/rudderstack/sdk/flutter/RudderSdkFlutterPlugin.java [hunk 4]
```diff
@@ -229,6 +230,12 @@ public RudderConfig getRudderConfig(Map<String, Object> configMap) {
         .withTrackLifecycleEvents((Boolean) configMap.get("trackLifecycleEvents"))
         .withRecordScreenViews((Boolean) configMap.get("recordScreenViews"))
         .withControlPlaneUrl((String) configMap.get("controlPlaneUrl"));
+    String dataResidencyServer = (String) configMap.get("dataResidency");
+    if(dataResidencyServer.equals("US")){
```
Comment (bardisg) [inline packages/plugins/rudder_plugin_android/android/src/main/java/com/rudderstack/sdk/flutter/RudderSdkFlutterPlugin.java:234]: I would say if it is UE ten set to EU else US. US is our default one

File: packages/plugins/rudder_plugin_interface/lib/src/constants.dart [hunk 5]
```diff
@@ -31,6 +31,8 @@ class Constants {
   // whether we should record screen views automatically
   static const bool RECORD_SCREEN_VIEWS = false;
 
+  static const String DATA_RESIDENCY_VALUE = "US";
```
Comment (bardisg) [inline packages/plugins/rudder_plugin_interface/lib/src/constants.dart:34]: i this a default value? if yes rename to DEFAULT_DATA_RESIDENCY or something

File: packages/plugins/rudder_plugin_interface/lib/src/models/rudder_config.dart [hunk 6]
```diff
@@ -119,6 +120,20 @@ class RudderConfig {
       _mobileConfigMap['controlPlaneUrl'] = controlPlaneUrl;
       _webConfigMap['configUrl'] = controlPlaneUrl;
     }
+    if (Utils.isEmpty(dataResidency)) {
+      RudderLogger.logError(
+          "dataResidency can not be null or empty. Set to default.");
+      _mobileConfigMap['dataResidency'] = Constants.DATA_RESIDENCY_VALUE;
+      _webConfigMap['dataResidency'] = Constants.DATA_RESIDENCY_VALUE;
+    } else if (dataResidency != "US" && dataResidency != "EU") {
```
Comment (bardisg) [inline packages/plugins/rudder_plugin_interface/lib/src/models/rudder_config.dart:128]: US and EU can be constant values

Comment (desusai7) [reply packages/plugins/rudder_plugin_interface/lib/src/models/rudder_config.dart:128]: yes, this was the minor improvement I was talking about

File: packages/plugins/rudder_plugin_interface/lib/src/constants.dart (hunk 5 above)
Comment (desusai7) [reply packages/plugins/rudder_plugin_interface/lib/src/constants.dart:34]: sure, will change it

File: packages/example/ios/Podfile.lock (hunk 1 above)
Comment (desusai7) [reply packages/example/ios/Podfile.lock:36]: it is not the latest after release, will update it

File: packages/example/ios/Podfile.lock (hunk 2 above)
Comment (desusai7) [reply packages/example/ios/Podfile.lock:246]: I guess, this has been removed by mistake, will rectify it

File: packages/plugins/rudder_plugin_android/android/build.gradle (hunk 3 above)
Comment (desusai7) [reply packages/plugins/rudder_plugin_android/android/build.gradle:41]: it is not the latest after release, will update it

File: packages/plugins/rudder_plugin_android/android/src/main/java/com/rudderstack/sdk/flutter/RudderSdkFlutterPlugin.java (hunk 4 above)
Comment (desusai7) [reply packages/plugins/rudder_plugin_android/android/src/main/java/com/rudderstack/sdk/flutter/RudderSdkFlutterPlugin.java:234]: Sure, will update it

File: packages/plugins/rudder_plugin_interface/lib/src/models/rudder_config.dart (hunk 6 above)
Comment (desusai7) [reply packages/plugins/rudder_plugin_interface/lib/src/models/rudder_config.dart]: yes, changed them to enums

File: packages/plugins/rudder_plugin_interface/lib/src/constants.dart (hunk 5 above)
Comment (desusai7) [reply packages/plugins/rudder_plugin_interface/lib/src/constants.dart]: updated it

File: packages/example/ios/Podfile.lock (hunk 2 above)
Comment (desusai7) [reply packages/example/ios/Podfile.lock:246]: Rectified it

File: packages/plugins/rudder_plugin_android/android/build.gradle (hunk 3 above)
Comment (desusai7) [reply packages/plugins/rudder_plugin_android/android/build.gradle]: updated this

File: packages/example/ios/Podfile.lock (hunk 1 above)
Comment (desusai7) [reply packages/example/ios/Podfile.lock]: updated to the latest

**Comment by desusai7** [comment]
- Updated dart SDK version constraint to >=2.17.6 across all packages to use enhanced enums (we will need to support dart version > 3 as well in future
- Removed un-used dependencies across all packages
- Added Unit Tests Cases for the Data Residency Feature
- Upgraded to the latest version of the underlying Native SDKs

**Review by bardisg — APPROVED** [review]

**Review by pallabmaiti — APPROVED** [review]

**Pull request merged**